pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -519,35 +519,42 @@ public void sendAsync(Message<?> message, SendCallback callback) {
 
         // chunked message also sent individually so, try to acquire send-permits
         for (int i = 0; i < (totalChunks - 1); i++) {
-            if (!canEnqueueRequest(callback, message.getSequenceId(), 0 /* The memory was already reserved */)) {
+            if (!conf.isBlockIfQueueFull() && !canEnqueueRequest(callback, message.getSequenceId(),
+                    0 /* The memory was already reserved */)) {
                 client.getMemoryLimitController().releaseMemory(uncompressedSize);
                 semaphoreRelease(i + 1);
                 return;
             }
         }
```

Comment:
The root cause is that for a chunked message with N chunks, since chunks are sent asynchronously one by one in the same thread, the time order will be:

1. Chunk 0 before send: Acquire 1 semaphore and `size0` memory.
2. Chunk 0 after send: Release 1 semaphore and `size0` memory.
3. Chunk 1 before send: Acquire 1 semaphore and `size1` memory.
4. Chunk 1 after send: Release 1 semaphore and `size1` memory. 
5. ...

In this case, we don't need N spots from the semaphore and `sum(size0, size1, ...)` bytes from the memory limiter. Instead, we only need 1 spot from the semaphore and `max(size0, size1, ...)` bytes from the memory limiter.

Therefore, we should not call `canEnqueueRequest` for chunk 1 and later, no matter if BlockIfQueueFUll is enabled. Instead, we should call `canEnqueueRequest` in each iteration in the loop.

Reply:
I think it's a great point to make the implementation simple and easy to understand.

Reply:
@BewareMyPower Yes. Chunks are sent asynchronously one by one in the same thread. But from your example, it seems that they are sent synchronously one by one. My understanding is that there may be a case like this in your approach:

Chunk 0 before send: Acquire 1 semaphore and max(size...) memory
Chunk 1 before send: 
// Here we already occupy 2 semaphores and (size0+size1) memory. It's not the correct behavior.
Chunk 0 after send: 
Chunk 1 after send: Release 1 semaphore and max(size...) memory.
...

Please correct if I'm wrong.

Reply:
Oh I see. Mark it as resolved.